smtpcode.c

```diff
@@ -0,0 +1,60 @@
+#include "smtpcode.h"
+
+#include "substdio.h"
+#include "stralloc.h"
+
+#define HUGESMTPTEXT 5000
+
+extern stralloc smtptext;
+extern substdio smtpfrom;
+
+static void get(unsigned char *uc)
+{
+  char *ch = (char *)uc;
+  substdio_get(&smtpfrom,ch,1);
+  if (*ch != '\r')
+    if (smtptext.len < HUGESMTPTEXT)
+     if (!stralloc_append(&smtptext,ch)) temp_nomem();
+}
+
+static int get_digit(unsigned char *uc)
+{
+  get(uc);
+  if (*uc < '0' || *uc > '9')
+    return 1;
+  return 0;
+}
+
+unsigned long smtpcode()
+{
+  unsigned char ch;
+  unsigned long code;
+  int err = 0;
+
+  if (!stralloc_copys(&smtptext,"")) temp_nomem();
+
+  err += get_digit(&ch); code = ch - '0';
+  /* valid SMTP codes are >= 200 */
+  if (ch < '2')
+    err = 1;
```

Comment:
In theory one could send so much crap that it will eventually overflow and pass as good, but why would one send 2^32 errors just to fool us to think the mail was accepted?

Reply:
This would happen if the input is closed during the loop, after 200 for instance for `200 `<EOF>?

But the check on the content would prevent the loop to go till the end.

replacing for(;;) by while(err == 0) would cut the superstition.

Reply:
But that would break that this loop reads in all lines of a multiline reply.